statistics/cmsketch.go

```diff
@@ -628,16 +629,110 @@ func NewTopN(n int) *TopN {
+//     1. `*TopN` is the final global-level topN.
+//     2. `[]TopNMeta` is the left topN value from the partition-level TopNs, but is not placed to global-level TopN. We should put them back to histogram latter.
+//     3. `[]*Histogram` are the partition-level histograms which just delete some values when we merge the global-level topN.
+func MergePartTopN2GlobalTopN(sc *stmtctx.StatementContext, topNs []*TopN, n uint32, hists []*Histogram, isIndex bool) (*TopN, []TopNMeta, []*Histogram, error) {
+	if checkEmptyTopNs(topNs) {
+		return nil, nil, hists, nil
+	}
+
+	partNum := len(topNs)
+	topNsNum := make([]int, partNum)
+	removeVals := make([][]TopNMeta, partNum)
+	for i, topN := range topNs {
+		if topN == nil {
+			topNsNum[i] = 0
+			continue
+		}
+		topNsNum[i] = len(topN.TopN)
+	}
+	// Different TopN structures may hold the same value, we have to merge them.
+	counter := make(map[hack.MutableString]float64)
+	// datumMap is used to store the mapping from the string type to datum type.
+	// The datum is used to find the value in the histogram.
+	datumMap := make(map[hack.MutableString]types.Datum)
+	for i, topN := range topNs {
+		if topN.TotalCount() == 0 {
+			continue
+		}
+		for _, val := range topN.TopN {
+			encodedVal := hack.String(val.Encoded)
+			counter[encodedVal] += float64(val.Count)
+			// We need to check whether the value corresponding to encodedVal is contained in other partition-level stats.
+			// 1. Check the topN first.
+			// 2. If the topN doesn't contain the value corresponding to encodedVal. We should check the histogram.
+			for j := 0; j < partNum; j++ {
+				if j == i {
+					continue
+				}
+				if topNs[j].findTopN(val.Encoded) == -1 {
+					// Get the encodedVal from the hists[j]
+					datum, exists := datumMap[encodedVal]
+					if !exists {
+						// If the datumMap does not have the encodedVal datum,
+						// we should generate the datum based on the encoded value.
+						// This part is copied from the function MergePartitionHist2GlobalHist.
+						var d types.Datum
+						if isIndex {
+							d.SetBytes(val.Encoded)
+						} else {
+							var err error
+							if types.IsTypeTime(hists[0].Tp.Tp) {
+								// handle datetime values specially since they are encoded to int and we'll get int values if using DecodeOne.
+								_, d, err = codec.DecodeAsDateTime(val.Encoded, hists[0].Tp.Tp, sc.TimeZone)
+							} else {
+								_, d, err = codec.DecodeOne(val.Encoded)
+							}
+							if err != nil {
+								return nil, nil, nil, err
+							}
+						}
+						datumMap[encodedVal] = d
+						datum = d
+					}
+					// Get the row count which the value is equal to the encodedVal from histogram.
+					count := hists[j].equalRowCount(datum, isIndex)
+					if count != 0 {
+						counter[encodedVal] += count
+						// Remove the value corresponding to encodedVal from the histogram.
+						removeVals[j] = append(removeVals[j], TopNMeta{Encoded: val.Encoded, Count: uint64(count)})
+					}
+				}
+			}
+		}
```

Comment:
How about splitting this logic(loop) into two parts? We can accumulate all part-level topn into `counter` in the first part and adjust all values in `counter` by `Hist` in the second part. Is it more clear? How do you think?